### ISSUE TITLE
Set the gradients of pruned filters zero

### ISSUE THREAD

**Issue by xingyueye** [he-y/filter-pruning-geometric-median#10]
@he-y thanks for your work and sharing!
I have a question: In train() function:
# Mask grad for iteration
        m.do_grad_mask()
        optimizer.step()
How do you change model's grad through do_grad_mask for m?

**Comment by he-y** [comment]
Please refer to [this issue](https://github.com/he-y/filter-pruning-geometric-median/issues/9) for details.

**Comment by xingyueye** [comment]
thanks. By the way, i think it could set grad zero through set bn.weigths and bn.bias to zero in do_mask phase.
And i'm confused about the following code (pruning_imagenet.py: line 525-528):
            indices = torch.LongTensor(filter_large_index).cuda()
            weight_vec_after_norm = torch.index_select(weight_vec, 0, indices).cpu().numpy()
            # for euclidean distance
            similar_matrix = distance.cdist(weight_vec_after_norm, weight_vec_after_norm, 'euclidean')
Why do you need to sort the filters based on norm first and then calculate the distance? Actually, I find the similar_matrix index is similar to indices, that is to say, the results obtained by the two methods are similar.So can we think that the channel with smaller norm is the channel of Geometric Median as well?
Hope for your reply, thanks ^^

**Comment by xingyueye** [comment]
@he-y thanks. By the way, i think it could set grad zero through set bn.weigths and bn.bias to zero in do_mask phase.
And i'm confused about the following code (pruning_imagenet.py: line 525-528):
```
    indices = torch.LongTensor(filter_large_index).cuda()
    weight_vec_after_norm = torch.index_select(weight_vec, 0, indices).cpu().numpy()
    # for euclidean distance
    similar_matrix = distance.cdist(weight_vec_after_norm, weight_vec_after_norm, 'euclidean')
```
Why do you need to sort the filters based on norm first and then calculate the distance? Actually, I find the similar_matrix index is similar to indices, that is to say, the results obtained by the two methods are similar.So can we think that the filters with smaller norm is the filters of Geometric Median as well?
Hope for your reply, thanks ^^

**Comment by he-y** [comment]
@xingyueye 
Q1: Why do you need to sort the filters based on norm first and then calculate the distance? 
A1: This operation is for the mix of norm-based criterion and FPGM. 
Quote from our paper:
>  paper: “FPGM-only 40%” means 40% filters of the layer are selected with FPGM only, while  “FPGM-mix  40%”  means  30% filters of the layer are selected with  FPGM,  and the remaining 10%  filters are selected with norm-based criterion.

Q2: So can we think that the filters with smaller norm is the filters of Geometric Median as well?
A2: No, they are different. You can set the [pruning ratio of these two criteria](https://github.com/he-y/filter-pruning-geometric-median/blob/master/pruning_imagenet.py#L52-L53) to activate only one criterion.

**Comment by xingyueye** [comment]
@he-y I get it. I am reading your article these days, it is very helpful to me, thank you very much.

**Comment by he-y** [comment]
@xingyueye Thanks for your interests.

**Comment by meijieru** [comment]
@he-y Since the masked filters are set to zero, and the gradient of them is masked, are they won't involve the training/finetuning process permanently?
And why do you re-prune after every epoch?

**Comment by he-y** [comment]
@meijieru 

>  Since the masked filters are set to zero, and the gradient of them is masked, are they won't involve the training/finetuning process permanently?

Sorry, I don't understand your question.

> And why do you re-prune after every epoch?

It's a hyper-parameter. Please see section 4.4 for detailed analysis.

**Comment by meijieru** [comment]
I mean once filters are pruned, they will be set to zero. And since you set the gradient of pruned filter to zero, they won't be updated during training. So will they be pruned again in next pruning?
If so, why should we have next pruning?

**Comment by he-y** [comment]
@meijieru 
I have tried the iteration level and epoch level [gradient mask](https://github.com/he-y/filter-pruning-geometric-median/blob/master/pruning_imagenet.py#L281).
This operation would keep both setting right and do not influence the results, so I keep it.

**Comment by meijieru** [comment]
Got it. So if we mask the gradient, it won't influence the result.
Thanks!
